### PR TITLE
amd-hybrid-graphics: fix race condition

### PR DESCRIPTION
**Pull request by lheckemann**
###### Motivation for this change
#19775 

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I'm really not sure about the correctness of my implementation. Feedback is very welcome!

### CODE REVIEW

File: nixos/modules/services/hardware/amd-hybrid-graphics.nix [hunk 1]
```diff
@@ -25,15 +25,22 @@
...
-      wantedBy = [ "multi-user.target" ];
+      before = [ "systemd-vconsole-setup.service" "display-manager.service" ];
+      requires = [ "sys-kernel-debug.mount" "vgaswitcheroo.path" ];
       serviceConfig = {
         Type = "oneshot";
         RemainAfterExit = true;
-        ExecStart = "${pkgs.bash}/bin/sh -c 'echo -e \"IGD\\nOFF\" > /sys/kernel/debug/vgaswitcheroo/switch; exit 0'";
-        ExecStop = "${pkgs.bash}/bin/sh -c 'echo ON >/sys/kernel/debug/vgaswitcheroo/switch; exit 0'";
+        ExecStart = "${pkgs.bash}/bin/sh -c 'echo -e \"IGD\\nOFF\" > /sys/kernel/debug/vgaswitcheroo/switch'";
+        ExecStop = "${pkgs.bash}/bin/sh -c 'echo ON >/sys/kernel/debug/vgaswitcheroo/switch'";
```
Comment (grahamc) [inline nixos/modules/services/hardware/amd-hybrid-graphics.nix]: Do you know why the exit 0's were there? / Why did you remove them?

**Comment by lheckemann** [comment]
I removed them because they hid the fact that it was failing, which made the actual problem less evident. I don't know why they were there.

**Comment by grahamc** [comment]
LGTM, @wkennington can you take a look? OR someone with this hardware?

File: nixos/modules/services/hardware/amd-hybrid-graphics.nix [hunk 2]
```diff
@@ -25,15 +25,22 @@
...
       };
     };
+    systemd.paths."vgaswitcheroo" = {
```
Comment (joachifm) [inline nixos/modules/services/hardware/amd-hybrid-graphics.nix]: Can you elaborate on the motivation for this versus using `unitConfig.ConditionPathExists` in the above unit?

File: nixos/modules/services/hardware/amd-hybrid-graphics.nix [hunk 3]
```diff
@@ -25,15 +25,22 @@
       path = [ pkgs.bash ];
       description = "Disable AMD Card";
       after = [ "sys-kernel-debug.mount" ];
-      requires = [ "sys-kernel-debug.mount" ];
-      wantedBy = [ "multi-user.target" ];
+      before = [ "systemd-vconsole-setup.service" "display-manager.service" ];
```
Comment (joachifm) [inline nixos/modules/services/hardware/amd-hybrid-graphics.nix]: Hm, I'm a litt confused about the dependency graph here; previously this unit would be pulled in by the default target, but what pulls it in now?

Comment (joachifm) [inline nixos/modules/services/hardware/amd-hybrid-graphics.nix]: Oh, nevermind; I see it's supposed to be pulled in by the path unit below :)

File: nixos/modules/services/hardware/amd-hybrid-graphics.nix (hunk 2 above)
Comment (lheckemann) [inline nixos/modules/services/hardware/amd-hybrid-graphics.nix]: Sorry, missed this — we want to start the unit *once the path appears*, not *if it exists*.

File: nixos/modules/services/hardware/amd-hybrid-graphics.nix (hunk 1 above)
Comment (lheckemann) [inline nixos/modules/services/hardware/amd-hybrid-graphics.nix]: #19775 — they caused the script to fail silently. I don't know why they were there.

**Comment by lheckemann** [comment]
Since nobody else seems to be using this module and able to review the change, perhaps it would make sense to remove the module entirely? Nobody else has complained about it being broken, so either nobody else has encountered the race condition (seems unlikely to me) or nobody else is using it and it would be better off living only in my personal config.
